README
FACEBOOK PROJECT

Il progetto è sviluppato in java usando l'ambiente eclipse, per ora sto sviluppando 
l'implementazione della chat, risolto il problema con i cookies per il login.

Nella cartella *python* è presente l'implementazione in python della chat
funzionante, può essere usata per studiare il protocollo di facebook e per fare
dei test.


CHANGELOG
30/11/09:
-BasicApi:
	- implementato il recuper del channel nella getInfos();
	- aggiunto un semplice controllo degli errori nella getInfos();

-Urls:
	- aggiunte le stringhe per la ricerca del post_form_id e del 
	  channel;

27/11/09:
-BasicApi:
	- implementato il recupero dell'user_id nella getInfos();
	- implementato il recupero del post_form_id nella getInfos();

21/11/09:
-BasicApi:
	- risolto il problema dei cookies, login effettuato con successo;
	- aggiunta classe "src/utils/MyCookieManager.java" per gestire i cookies;


src/test/Login.java
package test;

import java.net.MalformedURLException;

import fb.*;

public class Login
{

	/**
	 * effettua il login, utilizzo login "email" "password"
	 * @param args : "email" "password"
	 */
	public static void main (String[] args)
	{
		try
		{
			Api fb = new BasicApi (args[0], args[1]);
			
		} 
		catch (MalformedURLException e)
		{
			e.printStackTrace();
		}
		catch (Exception e)
		{
			System.out.println ("Eccezzione inaspettata");
			e.printStackTrace();
		}

	}

}


src/fb/Api.java
package fb;

/**
 * Rappresenta lo scheletro delle api, tutti metodi modificano variabili
 * interne
 * @author pausa
 *
 */
public interface Api
{
	/**
	 * Effettua la procedura di login
	 * @param uname : stringa che indica l'user name
	 * @param password : stringa che indica la password
	 */
	void login (String uname, String password);
	
	/**
	 * recupera id utente, post_form_id e channel
	 */
	void getInfos ();
	
	/**
	 * ottiene la buddyList
	 * @return: da definire, appena avrò deciso l'oggetto migliore...
	 */
	void getBuddyList ();
	
	/**
	 * Manda un messaggio
	 * @param to : stringa contenente l'id del destinatario
	 * @param text : testo del messaggio
	 */
	void send (String to, String text);
	
	/**
	 * riceve un messaggio dal server
	 * @return: da definire
	 */
	void receive ();
}


src/fb/Urls.java
package fb;

/**
 * contiene gli url necessari per l'interrogazione a facebook, ogni url
 * viene costruito a partire da FB_BASE
 * @author pausa
 *
 */

public abstract class Urls
{
	public final static String FB_BASE = new String 
										("http://www.facebook.com");
	
	public final static String FB_LOGIN_PAGE = new String ("/login.php");
	
	public final static String FB_HOME_PAGE = new String ("/home.php");
	
	public final static String FB_BUDDY_UPDATE = new String 
											("/ajax/presence/update.php");
	
	public final static String FB_SEND = new String 
												("/ajax/chat/send.php");
	
	public final static String FB_CHAT_WIN = new String 
												("/presence/popout.php");
	
	public final static String USER_AGENT = new String
								("Mozilla/5.0 (X11; U; Linux i686; en-US; " +
										"rv:1.9.1.3) Gecko/20090913 " +
										"Firefox/3.5.2");
	
	public final static String FB_POST_FORM = new String ("<input type=" +
									 	"\"hidden\" id=\"post_form_id\" " +
										   "name=\"post_form_id\" value=");
	
	public final static String FB_CHANNEL = new String ("channelManager.iframeLoad");
}


python/api/FBMain.py
# -*- coding: utf-8 -*-

import FB
import threading
import sys

class BuddyList ():

    def __init__ (self, name, password):
        self.fb = FB.FBHandler (name, password)
        self.updateBuddyList()
        self.showBuddyList()

        #inizializza il thread per la ricezione dei messaggi
        self.receive = receiveRoutine (self.fb)

    def updateBuddyList(self):
        b_array = []
        res = self.fb.buddy()

        if res["error"] != 0:
            print "error: " + str(res)
            return [("-1", ">> errore <<")]

        buddies = res["payload"]["buddy_list"]["userInfos"]
        
        for x,y in buddies.items():
            b_array.append ((x , y["name"]))

        self.buddyList = b_array

        return

        
    def showBuddyList (self):
        i = 0

        for x in self.buddyList:
            print str(i) + ": " + x[1]
            i+=1

        return

    def send (self, index, txt):
        res = self.fb.send(self.buddyList[index][0], txt)
        
        if res["error"] != 0:
            print ">> Errore nell'invio del messaggio: " + res["errorDescription"]
            print "\tcodice: " + str(res["error"])

        return

    def exitChat (self):
        sys.exit()

    #esegue il thread per la ricezione dei messaggi
    def startReceiveRoutine(self):
        self.receive.start()
        return


class receiveRoutine (threading.Thread):

    def __init__ (self, fb):
        threading.Thread.__init__(self)
        self.fb = fb

    def run (self):
        log = 0
        error = False

        while (True):
            res = {"t": "continue"}
            while (res["t"] == "continue"):
                    res = self.fb.receive()

            if res ["t"] == "msg":
                msg = res["ms"][0]

                #gestione dei tipi di messaggi:
                #messaggio ricevuto
                if msg["type"] == "msg":
                    print msg["from_name"] + " -> " + msg["to_name"] + ":"
                    print "\t" + msg["msg"]["text"]
                    print ""

                #un contatto sta scrivendo
                elif msg["type"] == "typ":
                    pass

                #operazioni eseguite nella finestra di facebook
                elif msg["type"] == "feedpub":
                    pass

                elif msg["type"] == "close_chat":
                    pass

                elif msg["type"] == "focus_chat":
                    pass

                elif msg["type"] == "unfocus_chat":
                    pass

                #messaggi delle notifiche
                elif msg["type"] == "notification":
                    pass

                elif msg["type"] == "notifications_read":
                    pass

                #messaggio da un'applicazione
                elif msg["type"] == "app_msg":
                    pass

                #codice imprevisto
                else:
                    error = True

            elif res["t"] == "refresh":
                pass

            else:
                error = True

            if error == True:
                print "Errore di ricezione: " + str(res)
                s = str(log) + ":\n\t" + str(res) + "\n\n"
                f = open ("debug", "a")
                f.write(s)
                f.close()
                error = False
        

src/fb/BasicApi.java
package fb;

import java.io.*;
import java.net.*;
import java.util.*;

/**
 * api a basso livello, si occupa di comunicare direttamente con facebook
 * @author pausa
 *
 */

public class BasicApi implements Api
{
	private String user = null;				//id utente
	private String post_form_id = null;		//post_form_id
	private String channel = null;			//canale assegnato all'utente
	private int seq = -1;					//# di sequenza del messaggio
	private URL basePage = null;			//url della pagina base
	private CookieHandler cm = null;
	
	/**
	 * costruttore base, inizializza solo l'oggetto
	 * 
	 * @throws MalformedURLException nel caso in cui l'url sia errato
	 */
	public BasicApi () throws MalformedURLException
	{
		
		user = new String ("");
		post_form_id = new String ("");
		channel = new String ("");
		basePage = new URL (Urls.FB_BASE);
		
		initCookieManager();
		
	}
	
	/**
	 * costruisce l'oggetto effettuando il login e recuperando i dati
	 * 
	 * @param uname: stringa rappresentante il nome utente
	 * @param password: stringa rappresentante la password
	 * @throws MalformedURLException nel caso in cui l'url sia errato
	 */
	public BasicApi (String uname, String password) 
								throws MalformedURLException
	{
		this();
		this.login(uname, password);
		this.getInfos();
		
	}
	
	@Override
	public void login (String uname, String password)
	{
		try
		{	
			OutputStreamWriter out = null;
			URL login = null;
			HttpURLConnection loginConn = null;
			
			System.out.println("--------------LOGIN:----------------");
			
			//costruisco l'URL per il login
			login = new URL (basePage, Urls.FB_LOGIN_PAGE);
			
			HttpURLConnection.setFollowRedirects(false);
			
			//apro una connessione verso la pagina
			loginConn = openConnection(login);
			
			System.out.println (loginConn.getResponseCode());
			
			//ottengo i cookies
			loginConn.getContent();
			loginConn.disconnect();
			
			//stampa a video di prova
			printCookies(loginConn);

			//apro una connessione per fare la post request
			loginConn = openConnection(login);
			loginConn.setRequestMethod("POST");

			loginConn.setDoOutput(true);
			
			//inizializzo il buffer di scrittura
			out = new OutputStreamWriter(loginConn.getOutputStream());
			
			//lo riempio con i parametri per il login
			out.write("email=" + uname + "&pass=" + password);
			
			//lo spedisco a facebook
			out.flush();
			out.close();
			
			loginConn.disconnect();
			
			//verifico la risposta
			System.out.println (loginConn.getResponseCode());
			System.out.println (loginConn.getRequestMethod());	
			
			printCookies(loginConn);
			
			System.out.println("-------------------END LOGIN---------" +
														  "-------------");
						
		}
		catch (MalformedURLException e)
		{
			System.out.println ("error: " + e.getCause());
		}
		catch (UnknownServiceException e)
		{
			System.out.println ("la pagina non supporta il POST");
		}
		catch (IOException e)
		{
			System.out.println ("errore imprevisto");
			e.printStackTrace();
		}
	}
	
	@Override
	public void getInfos ()
	{
		try
		{
			System.out.println ("------------------------GET INFOS----------" +
																  "---------");
			//inizializzazione delle variabili
			URL infopage = null;
			HttpURLConnection infoConn = null;
			BufferedReader read = null;
			String line = null;
			
			int pfStart = -1, pfEnd = -1;
			CookieStore cs = ((CookieManager)cm).getCookieStore();
			
			//cerco l'user id all'interno dei cookies
			for (HttpCookie c : cs.getCookies())
			{
				if (c.getName().equals("c_user"))
				{
					user += c.getValue();
					break;
				}	
			}
			
			//stampo l'user id
			if (this.user.equals(""))
				System.out.println("Login errato");
			
			else 
				System.out.println("user id: " + user);
			
			//cerco il post form id all'interno della finestra di chat
			infopage = new URL(basePage, Urls.FB_CHAT_WIN);
			
			infoConn = openConnection(infopage);
			
			read = new BufferedReader(new InputStreamReader
											  (infoConn.getInputStream()));
			
			//cerco la linea corretta nel codice
			while ((line = read.readLine()) != null)
			{
				if (line.startsWith(Urls.FB_POST_FORM))
				{
					//estraggo la stringa
					pfStart = line.indexOf("value=");
					pfStart = line.indexOf("\"", pfStart);
					pfStart ++;
					
					pfEnd = line.indexOf("\"", pfStart);
					
					post_form_id = line.substring(pfStart, pfEnd);
					
				}
				else if (line.startsWith(Urls.FB_CHANNEL))
				{
					pfStart = line.lastIndexOf("channel");
					pfStart += 7;
					pfEnd = pfStart + 2;
					channel = line.substring(pfStart, pfEnd);	
				}
				
			}
			
			//stampo
			if (post_form_id == null)
				System.out.println("impossibile recuperare post_form_id");
			else
				System.out.println("post form: " + post_form_id);
			
			if (channel == null)
				System.out.println("impossibile recuperare channel");
			else
				System.out.println ("channel: " + channel);
			
			System.out.println ("----------------END GET INFOS--------------");
			
		}
		catch (MalformedURLException e)
		{
			System.out.println ("error: " + e.getCause());
		}
		catch (IOException e)
		{
			System.out.println ("errore imprevisto");
			e.printStackTrace();
		}
	}
	
	@Override
	public void getBuddyList ()
	{
		// TODO Auto-generated method stub
		
	}

	@Override
	public void receive ()
	{
		// TODO Auto-generated method stub
		
	}

	@Override
	public void send (String to, String text)
	{
		// TODO Auto-generated method stub
		
	}
	/**
	 * inizializza il CookieManager, non dovrebbe essere chiamato solo se 
	 * si vogliono azzerare i cookies
	 */
	private void initCookieManager ()
	{
		cm = new utils.MyCookieManager();
		java.net.CookieHandler.setDefault(cm);		
	}

	/**
	 * apre una nuova connessione verso l'url puntato da *u* dandogli 
	 * l'user-agent definito nel file Urls.java
	 * @param u
	 * @return
	 * @throws IOException
	 */
	private HttpURLConnection openConnection (URL u) throws IOException
	{
		HttpURLConnection conn = (HttpURLConnection)u.openConnection();
		setUserAgent(conn);
		return conn;	
	}
	
	/**
	 * imposta l'user agent per la connessione
	 * @param u
	 */
	private void setUserAgent (URLConnection u)
	{
		u.setRequestProperty("User-Agent", Urls.USER_AGENT);	
	}
	
	/**
	 * stampa a video i cookies restituiti dalla connessione
	 * @param conn
	 */
	private void printCookies (HttpURLConnection conn)
	{
		CookieStore cs = ((CookieManager)cm).getCookieStore();
		
		int i = 0;
		System.out.println ("----------Cookies--------------");
		for (i = 0; i < 99; i++)
		{
			String s = conn.getHeaderFieldKey(i);
			
			if (s != null && s.equals("Set-Cookie"))
			{
				System.out.println ("Set-Cookie: " + conn.getHeaderField(i));
			}
				
			
		}
		
		System.out.println("-------------------------------");
		
		for (HttpCookie c : cs.getCookies())
			System.out.println (c.toString());
		
		System.out.println("-------------------------------------------");
	}
	
}


src/utils/MyCookieManager.java
package utils;

import java.io.IOException;
import java.net.*;
import java.util.*;

/**
 * Classe sviluppata per modificare il comportamento del Cookie Manager di
 * default rispetto ad alcuni cookies.
 * @author pausa
 *
 */
public class MyCookieManager extends CookieManager
{

	public MyCookieManager ()
	{
		super();	
	}
	
	public MyCookieManager(CookieStore store,
            CookiePolicy cookiePolicy)
	{
		super (store, cookiePolicy);
	}
	
	/**
	 * modifico il metodo *put* in modo che rimuova il parametro *httponly*
	 * da ogni cookie.
	 */
	@Override
	public void put(URI uri, Map<String, List<String>> responseHeaders)
    													throws IOException
    {
		//inizializzo un nuovo header usare con il CookieManager originale
		Map <String, List<String>> newHeader = 
									new HashMap <String, List<String>>();
		
		List <String> temp = null, newList = new LinkedList <String>();
		
		//ottengo il valore associato alla chiave
		temp = responseHeaders.get("Set-Cookie");
		
		//se la chiave è presente
		if (temp != null)
		{
			/*aggiungo alla nuova lista tutti i cookie trovati, togliendo
			loro il parametro *httponly* */
			for (String s : temp)
				newList.add(s.replace("; httponly", ""));
			
			//inserisco in nuovi valori nel nuovo header
			newHeader.put("Set-Cookie", newList);
			
			//reinizializzo la lista di cookies
			newList = new LinkedList <String>();
		}
		
		
		temp = responseHeaders.get("Set-Cookie2");
		
		if (temp != null)
		{
			for (String s : temp)
				newList.add(s.replace("; httponly", ""));
			
			newHeader.put("Set-Cookie2", newList);
		}
		
		//chiamo il metodo *put* originale con il nuovo insieme di cookies
		super.put(uri, newHeader);
    }
}


python/api/FB.py
# -*- coding: utf-8 -*-
#################################
#                               #
#   API DI FACEBOOK IN PYTHON   #
#                               #
#################################

#inserimento dei moduli necessari
import time
import urllib2
import cookielib
import random
import re
import json
from urllib import urlencode
from HTMLParser import HTMLParser
from math import ceil

#ridenominazione dei metodi usati più di frequente
urlopen = urllib2.urlopen
Request = urllib2.Request
randrange = random.randrange

#ridefinizione dell'user-agent
AGENT =  {'User-Agent' : \
            'Mozilla/5.0 (X11; U; Linux i686; en-US; rv:1.9.1.3) Gecko/20090913 Firefox/3.5.2'}
                
#indirizzi usati
FB_BASE = "http://www.facebook.com"
FB_LOGIN_PAGE = FB_BASE + "/login.php"
FB_HOME_PAGE = FB_BASE + "/home.php"
FB_BUDDY_UPDATE = FB_BASE + "/ajax/presence/update.php"
FB_SEND = FB_BASE + "/ajax/chat/send.php"
FB_CHAT_WIN = FB_BASE + "/presence/popout.php"

#limiti dei valori random
R1 = 1000000000
R2 = 9999999999

#inizializzazione del gestore dei cookies
cj = cookielib.CookieJar()
opener = urllib2.build_opener (urllib2.HTTPCookieProcessor(cj))
urllib2.install_opener(opener)


class FBHandler (HTMLParser):
    user = -1
    post_form_id = -1
    channel = -1
    seq = -1

    def __init__ (self, email, password):
        HTMLParser.__init__(self)
        data = self.login (email, password)
        self.populate(data)
        self.receive()

    #piccolo parser HTML per recuperare user_id e post_form_id
    def handle_starttag (self, tag, attrs):
    
        #check = False
        #if tag == "input":
            #for x,y in attrs:
                #if x.lower() == "id" and y.lower() == "user":
                    #check = true
                
                #if check == true and x.lower() == "value":
                    #self.user = y
                    #print "numero utente: ", y
                    #break
            
        check = False    
        if tag == "input":
            for x,y in attrs:
                if x.lower() == "id" and y.lower() == "post_form_id":
                    check = True
                    
                if check == True and x.lower() == "value":
                    self.post_form_id = y
                    print "post_form_id: ", y
                    break
                    
    def findChannel (self, data):
        exp = re.compile ("channel\d+")
        s = exp.search(data)
        self.channel = s.group()[7:9]
        print "channel: ", self.channel
        
    def findUser (self, data):
        exp = re.compile ('window.presence = new Presence\("\d+"')
        s = exp.search(data)
        print ">> ", s.group()
        up = s.group().find('"')
        down = s.group().rfind('"')
        self.user = s.group()[(up+1) : down]
        print "uid: ", self.user
        
    def populate (self, data):
        self.findUser(data)
        self.findChannel(data)
        self.feed(data)
        
    #restituisce il corpo della homepage di facebook
    def getPage (self, url):

        #viene costruita la richiesta
        print ">> ottenendo la pagina di facebook"
        req = Request (url,"",AGENT)
        p = urlopen (req)
        
        #restituisce il corpo
        return p.read().decode("utf8")

    #procedura di login, restituisce il corpo della pagina principale di fb
    def login (self, email, password):

        print ">> inizio la procedura di login"
        #parametri necessari per il login
        data = {"email" : email, "pass" : password}
        
        print ">> ottengo la pagina di login"
        #viene prima richiesta la pagina
        p = urlopen (FB_LOGIN_PAGE)
        
        #si effettua l'autenticazione tramite una POST request
        req = Request (FB_LOGIN_PAGE, urlencode(data), AGENT)
        
        print ">> invio la richiesta"
        #si richiede la pagina, dovrebbe restituire la homepage di fb
        p = urlopen (req)
        
        if p.geturl() == FB_LOGIN_PAGE:
            print ">> problemi di login\n"
        
        print ">> login effettuato"
        #restituisce il corpo della homepage
        return self.getPage(FB_CHAT_WIN)
        
    #procedura per ottenere la buddy list
    def buddy (self):

        #parametri necessari per fare la richiesta
        data = {"buddy_list" : "1", "force_render" : "1", \
                    "post_form_id" : self.post_form_id, "user": self.user}
        
        #si costruisce la stringa per la POST request
        req = Request (FB_BUDDY_UPDATE, urlencode (data), AGENT)
        
        #si ottiene il risultato
        p = urlopen(req)
        
        #restituisce la lista contenente la buddy list
        return json.loads(p.read().decode("utf8")[9:])

    #procedura per inviare messaggi
    def send (self, receiver, text):

        #parametri necessari per la richiesta
        data = {"client_time" : str(int(ceil(time.time() * 1000))),\
                "msg_id" : str(randrange(R1, R2)), "msg_text" : text, "popped_out" : \
                "true", "post_form_id" : self.post_form_id, "to" : receiver}
        
        #si costruisce la stringa per la POST request
        req = Request (FB_SEND, urlencode (data), AGENT)
        
        #si ottiene il risultato
        p = urlopen(req)
        
        #restituisce l'esito della richiesta
        return json.loads(p.read().decode("utf8")[9:])
        
    def receive (self):
                 
        p = urlopen ("http://0.channel" + self.channel + \
                        ".facebook.com/x/" + str(randrange(R1,R2)) + \
                        "/false/p_" + self.user + "=" + str(self.seq))

        res = json.loads (p.read().decode("utf8")[9:])
        
        if res["t"] == "refresh":
            self.seq = res["seq"]
            
        elif res["t"] != "continue":
            self.seq += 1;
        
        return res
        
        